commit 52a666d7cb490595c1c13e687bb5802274467163 (refs/changes/75/2775/13)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2017-10-20T13:44:16-07:00 (2 years ago)
    
    joyent/node-cueball#132 Expose counters and simple metrics from CueBallConnectionPool
    Reviewed by: Cody Peter Mello <cody.mello@joyent.com>
    Reviewed by: Alex Wilson <alex.wilson@joyent.com>

diff --git a/CHANGES.adoc b/CHANGES.adoc
index f7394f2..a2ef7fb 100644
--- a/CHANGES.adoc
+++ b/CHANGES.adoc
@@ -6,6 +6,17 @@ toc::[]
 
 ## v2.x
 
+### v2.5.0
+
+New minor release, due to addition of new API.
+
+API Changes:
+
+ - The `Pool` class now has a method `getStats()`, which returns a snapshot
+   of the Pool instance's counters along with four additional numbers:
+   total connections, idle connections, partly-open connections, and the
+   number of requests queued on available connections (#132)
+
 ### v2.4.0
 
 New minor release, due to addition of new API.
diff --git a/docs/api.adoc b/docs/api.adoc
index e3bfe44..6a8d920 100644
--- a/docs/api.adoc
+++ b/docs/api.adoc
@@ -3,7 +3,7 @@
 :doctype: book
 :idprefix:
 :docinfo:
-
+:listing-caption: "Example: "
 :sectnums:
 # API
 
@@ -241,6 +241,26 @@ determining the root cause may require further analysis.
 
 Returns an Error object.
 
+### `#getStats()`
+
+Returns an object containing a snapshot of the current counters, along with
+with four additional numbers: total connections, idle connections, partly-open
+connections, and the number of requests queued waiting.
+
+.An object returned by #getStats()
+[source,js,caption="Example: "]
+-----------------------------------------------------------------------------
+{ counters:
+   { claim: 1,
+     'max-claim-queue': 1,
+     'queued-claim': 1 },
+  totalConnections: 1,
+  idleConnections: 1,
+  pendingConnections: 0,
+  waiterCount: 0
+}
+-----------------------------------------------------------------------------
+
 ## Resolver
 
 ### About the interface
diff --git a/lib/pool.js b/lib/pool.js
index 84cab63..e72d0b9 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -729,6 +729,26 @@ CueBallConnectionPool.prototype.printConnections = function () {
 	console.log('dead:', self.p_dead);
 };
 
+CueBallConnectionPool.prototype.getStats = function () {
+	var self = this;
+
+	// Get a snapshot of current counter values.
+	var counters = {};
+	Object.keys(this.p_counters).forEach(function (k) {
+		counters[k] = self.p_counters[k];
+	});
+
+	var stats = {
+		'counters': counters,
+		'totalConnections': Object.keys(self.p_connections).length,
+		'idleConnections': self.p_idleq.length,
+		'pendingConnections': self.p_initq.length,
+		'waiterCount': self.p_waiters.length
+	};
+
+	return (stats);
+};
+
 CueBallConnectionPool.prototype.claim = function (options, cb) {
 	var self = this;
 	var done = false;
diff --git a/package.json b/package.json
index e1bd03f..d6cdaf4 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "cueball",
-  "version": "2.4.0",
+  "version": "2.5.0",
   "description": "manage a pool of connections to a multi-node service where nodes are listed in DNS",
   "main": "lib/index.js",
   "dependencies": {
diff --git a/test/pool.test.js b/test/pool.test.js
index d2ecaaa..bda065d 100644
--- a/test/pool.test.js
+++ b/test/pool.test.js
@@ -899,6 +899,32 @@ mod_tape.test('cueball#111', function (t) {
 	});
 });
 
+mod_tape.test('cueball#132 getStats()', function (t) {
+	connections = [];
+	resolver = undefined;
+
+	recovery.default.retries = 2;
+	var pool = new mod_pool.ConnectionPool({
+		log: log,
+		domain: 'foobar',
+		spares: 2,
+		maximum: 2,
+		constructor: function (backend) {
+			return (new DummyConnection(backend));
+		},
+		recovery: recovery
+	});
+	var s = pool.getStats();
+	t.equal(typeof (s), 'object');
+	t.equal(Object.keys(s).length, 5);
+	t.equal(typeof (s['counters']), 'object');
+	t.equal(s['totalConnections'], 0);
+	t.equal(s['idleConnections'], 0);
+	t.equal(s['pendingConnections'], 0);
+	t.equal(s['waiterCount'], 0);
+	t.end();
+});
+
 mod_tape.test('cleanup sandbox', function (t) {
 	sandbox.restore();
 	t.end();
